### PR TITLE
New partitions

### PR DESCRIPTION
**Pull request by jdiaz4302**
This PR updates `01_process/src/C` to incorporate Jordan's suggestions about better train/test splits.

# Overview

Before - train/validation/test consisted of unique lakes across the entire date range. This likely allows some data leakage in the form of very nearby neighbors experiencing near-identical weather patterns and freeze/thaw responses. Split = 60/20/20.

orginial_ice_splits

Now - train/validation/test consist of unique years (1980-2001, 2002-2006, 2007-2015). The training set consists of lakes that do not appear in the validation/test sets; the validation and test sets share lakes but do not overlap temporally - even this discards a lot of data, so further exclusion was not chosen. Split = 41/9/7 (43% of data is discarded).

approximate_new_spatial_partitions  |  new_temporal_partitions

This split was modified to be exclusively train past/evaluate future based on one of the most promising candidate partition set that was generated from a random temporal interval search and identified via interactive plotting of train/evaluate partition sizes.

# Notes/caveats

I explored a lot of options - partitions consisting of continuous spatial regions vs mixed locations throughout the set; partitions consisting of continuous vs discontinuous vs hierarchical year intervals; combinations of the two. Requiring any form of exclusion seems to impose about 35-40% missing data (because most data occur in the same years and/or few lakes are monitored for the entire temporal range).

**Remember to review the `.md` folders for your own sanity**

# Review requested

Bug identification or design errors.

I have tried a lot of solutions to get around the data loss, but it seems futile. So, given that Jordan helped me iterate on this conceptually as the domain scientist and has agreed on it, I am focused on moving onto training to see if modeling results hold up with reduced and differently partitioned data.

### CODE REVIEW

File: 01_process/src/C_TrainTestSplits.md [hunk 1]
```diff
@@ -39,10 +49,18 @@ mapping_reference = "../in/MN_ice/raw_data_from_DNR/lake_ice_id_spreadsheet.xlsx
...
 date_format = '%Y-%m-%d'
 
 vars_to_keep = ['ShortWave', 'LongWave', 'AirTemp', 'RelHum', 'WindSpeed', 'Rain',
                 'Snow', 'ice', 'temp_0_x', 'MaxDepth', 'LakeArea']
+
+# need a decently long evaluation period from random search
+min_years_partition = 10
+
+# favorite output from the seed-set random search
+found_iteration = 262
```
Comment (jdiaz4302) [inline 01_process/src/C_TrainTestSplits.md:63]: This new config value denotes the picked partition in the random temporal interval search. This has been/will always be the same due to the random seed. If for some reason results change, the partition can be reidentified using the known evaluation years of 2002-2015

File: 01_process/src/C_TrainTestSplits.md [hunk 2]
```diff
@@ -209,112 +232,177 @@ DOW_proportions = DOW_proportions.rename(columns = {'DOW':'prop',
 DOW_proportions
 ```
 
-### Randomly sample lakes until we get 20% of the data
+```python
+df = df.merge(DOW_proportions, how = 'left', on = 'DOW')
```
Comment (jdiaz4302) [inline 01_process/src/C_TrainTestSplits.md:236]: Note that this left merge excludes any lakes with missing geographic information (n lakes = 2, n sequences = 3)

File: 01_process/src/C_TrainTestSplits.md [hunk 3]
```diff
@@ -209,112 +232,177 @@ DOW_proportions = DOW_proportions.rename(columns = {'DOW':'prop',
...
+```
+
+# Utility functions
 
 ```python
-def partition_by_lakes(dataset, possible_lakes, desired_prop):
```
Comment (jdiaz4302) [inline 01_process/src/C_TrainTestSplits.md:215]: this function that was previously used to gather unique lakes until a certain percent of the data set was represented is no longer used

File: 01_process/src/C_TrainTestSplits.md [hunk 4]
```diff
@@ -209,112 +232,177 @@ DOW_proportions = DOW_proportions.rename(columns = {'DOW':'prop',
...
+    
+    remaining_indices = np.asarray([not i for i in sub_indices])
+    
+    remaining_data = data[remaining_indices]
+    
+    return remaining_data
 ```
 
-# Very basic mapping of the partitions
+# Perform a random search of different year-bound partitions
```
Comment (jdiaz4302) [inline 01_process/src/C_TrainTestSplits.md:291]: The basic description of this algorithm is as follows -

1. Randomly select a year interval 
2. Ensure that year interval has at least 10 years (config value)
3. Set aside data within those years as evaluation. Data outside those years are train.
4. For both partitions, get the unique lakes and calculate what percent of the whole data set they represent
5. Identify the lakes that exclusively occur in the evaluation years
6. Calculate what percent of the whole data set is represented if evaluation years use only lakes unique to that interval and train years use the remaining lakes.
7. Repeat

File: 01_process/src/C_TrainTestSplits.md [hunk 5]
```diff
@@ -209,112 +232,177 @@ DOW_proportions = DOW_proportions.rename(columns = {'DOW':'prop',
...
+    dow_in_A_only = [dow for dow in DOW_proportions_A['DOW'] if dow not in DOW_proportions_B['DOW'].values]
+    dow_tracking.append(dow_in_A_only)
+    
+    # Track year bounds
+    parameters[k] = year_bounds_A
+    # Calculate how much data remains when partition A (bounded by year_bounds_A)
+    # uses only the lakes unique to its years and excludes those lakes
+    # from partition B (the training set)
+    partition_sizes[k] = [np.sum(DOW_proportions_B[DOW_proportions_B['DOW'].isin(dow_in_A_only) == False]['prop']),
+                          np.sum(DOW_proportions_A[DOW_proportions_A['DOW'].isin(dow_in_A_only)]['prop'])]
```
Comment (AndyMcAliley) [inline 01_process/src/C_TrainTestSplits.md:345]: Trying to understand these two lines. `partition_sizes` gets assigned two numbers in each of the 500 tries. The second number, which I think I understand, is the proportion of the data that are in lakes in partition A only. The first number is the fraction of data that are in lakes that are in partition B and are not only in partition A? I don't understand `.isin(dow_in_A_only) == False`. Shouldn't there be zero DOWs in partition B for which `dow_in_A_only` is true?

File: 01_process/src/C_TrainTestSplits.md [hunk 6]
```diff
@@ -209,112 +232,177 @@ DOW_proportions = DOW_proportions.rename(columns = {'DOW':'prop',
...
+eval_df = subset_by_year(eval_df, year_bounds_A)
+train_df = get_remaining_data(df, year_bounds_A)
+```
 
 ```python
-train_indices = []
-valid_indices = []
-test_indices = []
+# plot the evaluation and training data
+fig, ax = plt.subplots(1, 2, figsize = (12, 6))
```
[the remark this answers is not in the document]
Comment (AndyMcAliley) [reply 01_process/src/C_TrainTestSplits.md:373]: Nice clear plot!

**Review by AndyMcAliley — APPROVED** [review]
The algorithm to find splits is a neat approach. Thanks for sharing it! I left a few questions in the code. My main question is about the decision to have lakes appear in both the validation and test sets. Sorry if you've already given the answer: that there's no good way to partition without significant loss of data, and this was the compromise that you and Jordan settled on.

File: 01_process/src/C_TrainTestSplits.md [hunk 7]
```diff
@@ -15,14 +15,24 @@ jupyter:
...
 ```
 
+# What happens in this notebook?
+
+1. Additional static variables are added - lake area and lake depth
+2. Certain variables are removed (due to their missingness or their expected availability outside the study area)
+3. Data is split into train, validation, and test partitions
+  * The train partition consists of lakes and years that are not represented in the validation and test partitions
+    * Notablly, all training data consists of years that predate validation and test data (2001 and earlier)
+  * The validation and test partitions share lakes, but contain mutually exclusive years
```
[the remark this answers is not in the document]
Comment (jdiaz4302) [reply 01_process/src/C_TrainTestSplits.md:31]: Ideally test and validation set would have different lakes. The issue is just the data set is already small, then applying another mutually exclusive condition further shrinks the total N.

I don't think the idea is to apply the model to lakes that lack any data, rather to just makes the partitions as different as possible while maintaining some size to make sure evaluation does not have any sort of "soft" data leaks in the form of predicting for a test set lake after having seen an immediately neighboring lake go through the exact same weather conditions in the training set.

Jordan and I thought that having different years was more convincing than having different lakes when we need to only pick 1 exclusion.

**That being said...** if the validation set needs to share time or space with another partition, and we assume that not-sharing time is more convincing, maybe I should flip this so that the train set and validation set share lakes while the test set is entirely exclusive (rather than test set and validation set share lakes while the train set is entirely exclusive). 

I think that is better. Agree/disagree/thoughts, @AndyMcAliley ?

File: 01_process/src/C_TrainTestSplits.md (hunk 5 above)
Comment (jdiaz4302) [reply 01_process/src/C_TrainTestSplits.md:345]: Yes, you're right that this is redundant. 

This stemmed from some original implementation ideas where the lakes in partition A wouldn't *only* occur in partition A and would thus need removed from partition B

File: 01_process/src/C_TrainTestSplits.md (hunk 1 above)
Comment (jdiaz4302) [reply 01_process/src/C_TrainTestSplits.md:63]: Nice, I think I will adopt this. Documenting that the solution in https://stackoverflow.com/questions/22994423/difference-between-np-random-seed-and-np-random-randomstate was useful for knowing how to do this.

File: 01_process/src/C_TrainTestSplits.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (AndyMcAliley) [reply 01_process/src/C_TrainTestSplits.md:31]: I agree that it makes more sense to have the test set be entirely exclusive than for the validation and test sets to share lakes that aren't in the training set. 

One option would be to have two test sets: one for same lakes but different years, and one for different lakes and different years. That would allow you to test performance for lakes in the training set and outside of it. It seems like you could do that without further shrinking the total N because you're already partitioning by year - this is a way to make use of all data from the testing years.

File: 01_process/src/C_TrainTestSplits.md [hunk 8]
```diff
@@ -15,14 +15,24 @@ jupyter:
...
+import plotly.express as px
 ```
 
+# What happens in this notebook?
+
+1. Additional static variables are added - lake area and lake depth
+2. Certain variables are removed (due to their missingness or their expected availability outside the study area)
+3. Data is split into train, validation, and test partitions
+  * The train partition consists of lakes and years that are not represented in the validation and test partitions
+    * Notablly, all training data consists of years that predate validation and test data (2001 and earlier)
```
[the remark this answers is not in the document]
Comment (jdiaz4302) [reply 01_process/src/C_TrainTestSplits.md:30]: Adding to next commit for ease

File: 01_process/src/C_TrainTestSplits.md [hunk 9]
```diff
@@ -56,11 +55,16 @@ date_format = '%Y-%m-%d'
...
-min_years_partition = 10
+# compatible with earlier efforts, leaves enough to train and eval with
+# a 'soft' test set can be larger, consisting of train/eval lakes during test years
+# and test lakes during train/eval years
+test_set_prop = 0.08
 
-# favorite output from the seed-set random search
-found_iteration = 262
+test_start_year = 2010
+valid_start_year = 2005
```
Comment (jdiaz4302) [inline 01_process/src/C_TrainTestSplits.md:64]: Years adjusted to allow good amount of training data; keeping test set proportion approximately equivalent (marginally increased to PR's last iteration)

File: 01_process/src/C_TrainTestSplits.md [hunk 10]
```diff
@@ -236,173 +240,123 @@ DOW_proportions
 df = df.merge(DOW_proportions, how = 'left', on = 'DOW')
 ```
 
-# Utility functions
+# Perform a random search of different year-bound partitions
 
 ```python
-# get min and max from values that potentially have missing values
-def get_min_and_max(vals):
+def partition_by_lakes(dataset, possible_lakes, desired_prop, total_N, rng):
```
Comment (jdiaz4302) [inline 01_process/src/C_TrainTestSplits.md:246]: I actually brought this function back, but I am using it a little differently

File: 01_process/src/C_TrainTestSplits.md [hunk 11]
```diff
@@ -236,173 +240,123 @@ DOW_proportions
...
+    #     `dataset` is the dataset we're partitioning.
+    #        I assume it has a "DOW" column
+    #     `possible_lakes` are the DOWs that we sample.
+    #        As the sampling continues, we remove values.
+    #     `desired_prop` is the proportion of the data set
+    #        that we want this partition to cover.
+    #     `total_N` is the number of available sequences across
+    #        all lakes and years
+    #     `rng` is a np.random.RandomState object (opposed to
+    #        np.random.seed setting)
```
Comment (jdiaz4302) [inline 01_process/src/C_TrainTestSplits.md:258]: Implemented requested change from `np.random.seed()`

File: 01_process/src/C_TrainTestSplits.md (hunk 3 above)
Comment (jdiaz4302) [reply 01_process/src/C_TrainTestSplits.md:215]: Ignore, reimplemented upon other ideas

File: 01_process/src/C_TrainTestSplits.md (hunk 4 above)
Comment (jdiaz4302) [reply 01_process/src/C_TrainTestSplits.md]: This was reverted. The basic description of what happens now is as follows -

1. Based on everything we know now, a test set of ~8% leaves enough training data. Likewise, testing on future years is possible and desireable -> let's set aside 2010-2017 as testing years (8 years total; Jordan said 5 would be small, 8 makes sense and 10 would be great but acknowledges data loss)
2. Do lots of random attempts of the following...
3. Within those testing years, sample lakes until 8% of the sequences are represented
4. Exclude those test lakes from 1980-2009 data. Assign 1980-2004 as train and 2005-2009 as validation.
5. Keep track of the years and resulting number of testing, validation, and training sequences.
6. By design, all testing candidates represent 8% or more of the available data, so select the candidate test lakes which result in the most training data.

The end result is 50/16.5/8 split which is more training and validation data for the same testing data. Additionally, the testing data is most different. The remaining/discarded data could later be used as a 'soft' test set.

File: 01_process/src/C_TrainTestSplits.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (jdiaz4302) [reply 01_process/src/C_TrainTestSplits.md]: OK, I made these changes, and I think they were for the best - thanks, it helps us discard notably less data!

File: 01_process/src/C_TrainTestSplits.md [hunk 12]
```diff
@@ -428,6 +382,22 @@ save_partition_data(valid_df.index, valid_data_fpath)
 save_partition_data(test_df.index, test_data_fpath)
 ```
 
+<br><br><br><br>
+
+# Lake overlap checking
```
Comment (AndyMcAliley) [inline 01_process/src/C_TrainTestSplits.md:391]: Good idea!

File: 01_process/src/C_TrainTestSplits.md [hunk 13]
```diff
@@ -335,11 +381,27 @@ def save_partition_data(indices, fpath):
...
+
 ```python
+# Percent of valid lakes appearing in test set
+len([dow for dow in valid_df['DOW'].unique() if dow in test_df['DOW'].unique()]) / valid_df['DOW'].unique().shape[0]
+```
 
+```python
+# Percent of valid lakes appearing in test set
```
Comment (AndyMcAliley) [inline 01_process/src/C_TrainTestSplits.md:399]:
```suggestion
# Percent of valid lakes appearing in train set
```

**Comment by AndyMcAliley** [comment]
I skimmed through the changes, but didn't read in detail. From what I saw they look good! Would you like me to go through them closely?

**Comment by jdiaz4302** [comment]
I was leaving that up to your discretion, but I think the latest changes are more similar to the previous versions, so we can probably save your time. I'll probably merge it in tomorrow morning unless you have any objections

**Comment by AndyMcAliley** [comment]
> I'll probably merge it in tomorrow morning unless you have any objections
Looks good to me!
